zeebe/backup/src/test/java/io/camunda/zeebe/backup/management/ConcurrentBackupCompactionTest.java

```diff
@@ -0,0 +1,232 @@
+import io.atomix.raft.impl.LogCompactor;
+import io.atomix.raft.metrics.RaftServiceMetrics;
+import io.atomix.raft.partition.RaftPartition;
+import io.atomix.raft.storage.log.RaftLog;
+import io.atomix.utils.concurrent.ThreadContext;
+import io.camunda.zeebe.journal.JournalMetaStore;
+import io.camunda.zeebe.journal.Segments;
+import io.camunda.zeebe.journal.file.SegmentedJournal;
+import io.camunda.zeebe.scheduler.ActorScheduler;
+import io.camunda.zeebe.scheduler.SchedulingHints;
+import io.camunda.zeebe.snapshots.PersistedSnapshot;
+import io.camunda.zeebe.snapshots.impl.FileBasedSnapshotStore;
+import io.camunda.zeebe.snapshots.impl.FileBasedSnapshotStoreImpl;
+import io.camunda.zeebe.test.util.AutoCloseableRule;
+import io.camunda.zeebe.util.FileUtil;
+import io.camunda.zeebe.util.buffer.DirectBufferWriter;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.time.Duration;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import org.agrona.concurrent.UnsafeBuffer;
+import org.awaitility.Awaitility;
+import org.junit.Rule;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.api.io.TempDir;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.mockito.stubbing.Answer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@ExtendWith(MockitoExtension.class)
+public class ConcurrentBackupCompactionTest {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ConcurrentBackupCompactionTest.class);
+  private static ActorScheduler actorScheduler;
+  private static final String SNAPSHOT_FILE_NAME = "file1";
+  @Rule public final AutoCloseableRule autoclose = new AutoCloseableRule();
+  @TempDir Path dataDirectory;
+  private SegmentedJournal journal;
+  private FileBasedSnapshotStore snapshotStore;
+  private InMemoryMockBackupStore backupStore;
+  private BackupService backupService;
+  private final int nodeId = 1;
+  private final int partitionId = 1;
+  private LogCompactor logCompactor;
+  @Mock private ThreadContext threadContext;
+  @Mock private RaftLog raftLog;
+  @Mock private RaftServiceMetrics raftMetrics;
+
+  @BeforeAll
+  static void beforeAll() {
+    actorScheduler = ActorScheduler.newActorScheduler().build();
+    actorScheduler.start();
+  }
+
+  @AfterAll
+  static void afterAll() throws ExecutionException, InterruptedException {
+    actorScheduler.stop().get();
+  }
+
+  @BeforeEach
+  void setUp() {
+    // methods use by the compactor to schedule callbacks
+    doAnswer(
+            inv -> {
+              ((Runnable) inv.getArgument(0)).run();
```

Comment:
:question: So call backs are invoked immediately from the caller thread, inline? Why not just use a dummy `ThreadContext` which does that? We have a `DeterministicSingleThreadContext` for example, but you could also just create one:

```java
private static final class InlineThreadContext implements ThreadContext {

  @Override
  public Scheduled schedule(
      final Duration initialDelay, final Duration interval, final Runnable callback) {
    callback.run();
    return new Scheduled() {
      @Override
      public void cancel() {}

      @Override
      public boolean isDone() {
        return true;
      };
    };
  }

  @Override
  public void execute(final Runnable command) {
    command.run();
  }
}
```

Reply:
I'm not a big fan of mocks like this, I find it harder to understand what's going on than with stubs.

Reply:
I'll move it to `DeterministingSingleThreadContext`. :+1:

Reply:
`DeterministingSingleThreadContext` is defined in zeebe-atomix-cluster, I'll just use the stub you wrote.